File: litr-demo/src/main/java/com/linkedin/android/litr/demo/ExtractFramesFragment.java
/*
 * Copyright 2019 LinkedIn Corporation
 * All Rights Reserved.
 *
 * Licensed under the BSD 2-Clause License (the "License").  See License in the project root for
 * license information.
 */
package com.linkedin.android.litr.demo;

import android.graphics.Bitmap;
import android.graphics.Point;
import android.media.MediaMetadataRetriever;
import android.net.Uri;
import android.os.Bundle;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.linkedin.android.litr.codec.MediaCodecDecoder;
import com.linkedin.android.litr.demo.data.SourceMedia;
import com.linkedin.android.litr.demo.databinding.FragmentExtractFramesBinding;
import com.linkedin.android.litr.exception.MediaSourceException;
import com.linkedin.android.litr.filter.GlFilter;
import com.linkedin.android.litr.io.MediaExtractorMediaSource;
import com.linkedin.android.litr.io.MediaRange;
import com.linkedin.android.litr.io.MediaSource;
import com.linkedin.android.litr.io.MediaSourceFactory;
import com.linkedin.android.litr.render.GlThumbnailRenderer;
import com.linkedin.android.litr.thumbnails.ThumbnailExtractListener;
import com.linkedin.android.litr.thumbnails.ThumbnailExtractParameters;
import com.linkedin.android.litr.thumbnails.VideoThumbnailExtractor;
import com.linkedin.android.litr.utils.TranscoderUtils;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.Executors;

public class ExtractFramesFragment extends BaseTransformationFragment implements MediaPickerListener {

    private FragmentExtractFramesBinding binding;

    private ArrayAdapter<DemoFilter> filtersAdapter;

    private VideoThumbnailExtractor thumbnailExtractor;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        filtersAdapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_item, DemoFilter.values());
        filtersAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);

        thumbnailExtractor = new VideoThumbnailExtractor(
                requireContext().getApplicationContext(),
                Executors.newSingleThreadExecutor(),
                new ThumbnailExtractListener() {
                    @Override
                    public void onStarted(@NonNull String id, @NonNull List<Long> timestampsUs) {
                        List<Bitmap> bitmaps = new ArrayList<>(timestampsUs.size());
                        for (int i = 0; i < timestampsUs.size(); i++) {
                            bitmaps.add(null);
                        }
                        binding.filmStripTimeline.setFrameList(bitmaps);
                    }

                    @Override
                    public void onExtracted(@NotNull String id, int index, @Nullable Bitmap bitmap) {
                        binding.filmStripTimeline.setFrameAt(index, bitmap);
                    }

                    @Override
                    public void onCompleted(@NonNull String id) {

                    }

                    @Override
                    public void onCancelled(@NonNull String id) {

                    }

                    @Override
                    public void onError(@NonNull String id, @Nullable Throwable cause) {

                    }
                }, Looper.getMainLooper());
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        thumbnailExtractor.release();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentExtractFramesBinding.inflate(inflater, container, false);

        binding.setFragment(this);

        SourceMedia sourceMedia = new SourceMedia();
        binding.setSourceMedia(sourceMedia);

        binding.sectionPickVideo.buttonPickVideo.setOnClickListener(view -> pickVideo(ExtractFramesFragment.this));

        binding.spinnerFilters.setAdapter(filtersAdapter);
        binding.spinnerFilters.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                binding.setFilter(filtersAdapter.getItem(position).filter);
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        return binding.getRoot();
    }



    public void extractThumbnails(@NonNull SourceMedia sourceMedia, @Nullable GlFilter filter) {

        Point sourceSize = TranscoderUtils.getVideoDimensions(requireContext().getApplicationContext(), sourceMedia.uri);
        if (sourceSize == null) {
            return;
        }

        double durationSec = sourceMedia.duration;
        int timelineWidth = binding.filmStripTimeline.getWidth();
        int thumbHeight = binding.filmStripTimeline.getHeight();

        if (durationSec <= 0 || timelineWidth <= 0 || thumbHeight <= 0) {
            return;
        }
        int thumbCount = 10;
        double secPerThumbnail = durationSec / thumbCount;

        List<Long> timestamps = new ArrayList<>();
        for (int i = 0; i < thumbCount; i++) {
            long timestampUs = (long)(secPerThumbnail * 1_000_000.0) * i;
            timestamps.add(timestampUs);
        }

        MediaSourceFactory mediaSourceFactory = new MediaSourceFactory() {
            @NonNull
            @Override
            public MediaSource createMediaSource() throws MediaSourceException {
                return new MediaExtractorMediaSource(
                        requireContext().getApplicationContext(),
                sourceMedia.uri,
                        new MediaRange(0, Long.MAX_VALUE));
            }

            @NonNull
            @Override
            public MediaMetadataRetriever getRetriever() {
                MediaMetadataRetriever retriever = new MediaMetadataRetriever();
                retriever.setDataSource(requireContext(), sourceMedia.uri);
                return retriever;
            }
        };

        int thumbWidth = timelineWidth / thumbCount;

        Point destSize = new Point(thumbWidth, thumbHeight);

        thumbnailExtractor.extract(
                UUID.randomUUID().toString(), new ThumbnailExtractParameters(
                        mediaSourceFactory,
                        timestamps,
                        new MediaRange(0, Long.MAX_VALUE),
                        new MediaCodecDecoder(),
                        new MediaCodecDecoder(),
                        sourceSize,
                        destSize,
                        new GlThumbnailRenderer(Collections.singletonList(filter))
                )
        );
    }

    @Override
    public void onMediaPicked(@NonNull Uri uri) {
        SourceMedia sourceMedia = binding.getSourceMedia();
        updateSourceMedia(sourceMedia, uri);
    }

}

File: litr-demo/src/main/java/com/linkedin/android/litr/demo/ExtractFramesFragment.kt
/*
 * Copyright 2021 LinkedIn Corporation
 * All Rights Reserved.
 *
 * Licensed under the BSD 2-Clause License (the "License").  See License in the project root for
 * license information.
 */
package com.linkedin.android.litr.demo

import android.R
import android.graphics.Bitmap
import android.graphics.Point
import android.net.Uri
import android.os.Bundle
import android.util.LruCache
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.linkedin.android.litr.ExperimentalFrameExtractorApi
import com.linkedin.android.litr.demo.data.SourceMedia
import com.linkedin.android.litr.demo.databinding.FragmentExtractFramesBinding
import com.linkedin.android.litr.filter.GlFilter
import com.linkedin.android.litr.render.GlSingleFrameRenderer
import com.linkedin.android.litr.frameextract.FrameExtractMode
import com.linkedin.android.litr.frameextract.FrameExtractListener
import com.linkedin.android.litr.frameextract.FrameExtractParameters
import com.linkedin.android.litr.frameextract.VideoFrameExtractor
import java.io.ByteArrayOutputStream
import java.util.*

@OptIn(ExperimentalFrameExtractorApi::class)
class ExtractFramesFragment : BaseTransformationFragment(), MediaPickerListener {
    private lateinit var binding: FragmentExtractFramesBinding
    private lateinit var filtersAdapter: ArrayAdapter<DemoFilter>
    private lateinit var frameExtractor: VideoFrameExtractor
    private lateinit var framesAdapter: ExtractedFramesAdapter
    private lateinit var bitmapInMemoryCache: LruCache<Long, ByteArray>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val cacheSize = (Runtime.getRuntime().maxMemory() / 1024L) / 8L
        bitmapInMemoryCache = object : LruCache<Long, ByteArray>(cacheSize.toInt()) {
            override fun sizeOf(key: Long, byteArray: ByteArray): Int {
                return byteArray.size / 1024
            }
        }

        filtersAdapter = ArrayAdapter(requireContext(), R.layout.simple_spinner_item, DemoFilter.values()).apply {
            setDropDownViewResource(R.layout.simple_spinner_dropdown_item)
        }

        frameExtractor = VideoFrameExtractor(requireContext())
        framesAdapter = ExtractedFramesAdapter(frameExtractor, bitmapInMemoryCache)
    }

    override fun onDestroy() {
        super.onDestroy()
        frameExtractor.release()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        binding = FragmentExtractFramesBinding.inflate(inflater, container, false).also { binding ->
            binding.fragment = this@ExtractFramesFragment
            binding.sourceMedia = SourceMedia()
            binding.sectionPickVideo.buttonPickVideo.setOnClickListener { pickVideo(this@ExtractFramesFragment) }
            binding.spinnerFilters.adapter = filtersAdapter
            binding.spinnerFilters.onItemSelectedListener = object : OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                    filtersAdapter.getItem(position)?.filter?.let {
                        binding.filter = it
                    }
                }

                override fun onNothingSelected(parent: AdapterView<*>?) {}
            }
        }
        binding.framesRecycler.layoutManager = LinearLayoutManager(requireContext(), RecyclerView.HORIZONTAL, false)
        binding.framesRecycler.adapter = framesAdapter
        return binding.root
    }

    override fun onDestroyView() {
        binding.framesRecycler.adapter = null
        bitmapInMemoryCache.evictAll()
        super.onDestroyView()
    }

    fun extractThumbnails(sourceMedia: SourceMedia, filter: GlFilter?) {
        frameExtractor.stopAll()
        bitmapInMemoryCache.evictAll()

        val thumbCount = 50

        val videoDurationSec = sourceMedia.duration.toDouble()
        val thumbHeight = binding.framesRecycler.height

        if (videoDurationSec <= 0 || thumbHeight <= 0) {
            return
        }

        // Based on the video duration and number of thumbnails, obtain the list of timestamps (in microseconds),
        // of each frame we should extract.
        val secPerThumbnail = videoDurationSec / thumbCount
        val timestamps = (0 until thumbCount).map {
            (secPerThumbnail * 1000000.0).toLong() * it
        }

        // Because we want to apply the same filter to each frame, the renderer may be shared between all thumbnail requests.
        val renderer = GlSingleFrameRenderer(filter?.let { listOf(it) })

        // From each timestamp, construct the parameters to send to the thumbnail extractor.
        val params = timestamps.map {
            FrameExtractParameters(
                sourceMedia.uri,
                it,
                renderer,
                FrameExtractMode.Fast,
                Point(thumbHeight, thumbHeight),
                0L
            )
        }

        // Set the list of extraction params to the adapter. Note that frame extraction will only start when the adapter binds an item to a view.
        framesAdapter.loadData(params)

        // What follows is an optimization.
        // We request all the thumbnails with a low priority, and cache the resulting bitmaps in an in-memory cache.
        // Because of the lower priority, these frames will only load when the RecyclerView adapter is not requesting frames at a higher priority.
        // In a production application, an on-disk bitmap cache may be preferred.
        params.forEach {
            frameExtractor.extract(UUID.randomUUID().toString(), it.copy(priority = 100L), object: FrameExtractListener {
                override fun onExtracted(id: String, timestampUs: Long, bitmap: Bitmap) {
                    // Compress bitmap
                    val baos = ByteArrayOutputStream()
                    bitmap.compress(Bitmap.CompressFormat.JPEG, 90, baos)
                    val byteArray = baos.toByteArray()
                    // Add compressed bytes to in-memory cache
                    bitmapInMemoryCache.put(timestampUs, byteArray)
                }
            })
        }
    }

    override fun onMediaPicked(uri: Uri) {
        binding.sourceMedia?.let {
            updateSourceMedia(it, uri)
        }
    }
}
